core/src/main/java/org/apache/calcite/sql/dialect/MssqlSqlDialect.java

```diff
@@ -105,44 +104,53 @@ public MssqlSqlDialect(Context context) {
 
   @Override public void unparseCall(SqlWriter writer, SqlCall call,
       int leftPrec, int rightPrec) {
-    if (call.getOperator() == SqlStdOperatorTable.SUBSTRING) {
-      if (call.operandCount() != 3) {
-        throw new IllegalArgumentException("MSSQL SUBSTRING requires FROM and FOR arguments");
+    switch (call.getKind()) {
+    case FLOOR:
+      if (call.operandCount() != 2) {
+        super.unparseCall(writer, call, leftPrec, rightPrec);
+        return;
       }
-      SqlUtil.unparseFunctionSyntax(MSSQL_SUBSTRING, writer, call);
-    } else {
-      switch (call.getKind()) {
-      case FLOOR:
-        if (call.operandCount() != 2) {
-          super.unparseCall(writer, call, leftPrec, rightPrec);
-          return;
+      unparseFloor(writer, call);
+      break;
+    case TRIM:
+      unparseTrim(writer, call, leftPrec, rightPrec);
+      break;
+    case OTHER_FUNCTION:
+    case TRUNCATE:
+      unparseOtherFunction(writer, call, leftPrec, rightPrec);
+      break;
+    case CEIL:
+      final SqlWriter.Frame ceilFrame = writer.startFunCall("CEILING");
+      call.operand(0).unparse(writer, leftPrec, rightPrec);
+      writer.endFunCall(ceilFrame);
+      break;
+    case NVL:
+      SqlNode[] extractNodeOperands = new SqlNode[]{call.operand(0), call.operand(1)};
+      SqlCall sqlCall = new SqlBasicCall(ISNULL, extractNodeOperands,
+              SqlParserPos.ZERO);
+      unparseCall(writer, sqlCall, leftPrec, rightPrec);
+      break;
+    case EXTRACT:
+      unparseExtract(writer, call, leftPrec, rightPrec);
+      break;
+    case SUBSTRING:
+      final SqlWriter.Frame substringFrame = writer.startFunCall("SUBSTRING");
+      final int maxLength = 2147483647;
```

Comment:
Do not use an int of max supported length in MSSQL. Find the length of Operand(1) and use Len(operand(1)) - 1